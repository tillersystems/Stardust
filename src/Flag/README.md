# Flag

### Usage

```jsx
import { Flag } from '@tillersystems/stardust';
```

<!-- STORY -->

### Description

A Flag displays a country flag from the list in `data.js`. It is drawed with a svg so added data should be kept as light as possible.

<!-- PROPS -->

### Example

```jsx
import { Flag } from '@tillersystems/stardust';

render() {
  return <Flag name="fr" />;
}
```
