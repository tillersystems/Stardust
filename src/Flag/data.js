import React from 'react';

const es = () => (
  <>
    <path fill="#c60b1e" d="M32 88h448v336H32z" />
    <path fill="#ffc400" d="M31 181h450v150H31z" />
  </>
);

const fr = () => (
  <>
    <path fill="#fff" d="M32 88h448v336H32z" />
    <path fill="#00267f" d="M32.332 88h149.336v336H32.332z" />
    <path fill="#f31830" d="M330.663 88H480v336H330.663z" />
  </>
);

const gb = () => (
  <>
    <path fill="#006" d="M32 88h448v336H32z" />
    <path
      d="M32 88v37.566L429.914 424h50.086v-37.564L82.086 88H32zm448 0v37.565L82.086 424H32v-37.565L429.914 88h50.086z"
      fill="#FFF"
      fillRule="nonzero"
    />
    <path
      d="M218.667 88v336h74.667V88H218.667zM32 200v112h448V200H32z"
      fill="#FFF"
      fillRule="nonzero"
    />
    <path
      d="M32 222.4v67.2h448v-67.2H32zm201.6-134.4v336h44.799V88H233.6zM32 424l149.334-112h33.39L65.391 424H32zm0-336 149.334 112h-33.391L32 113.045V88zm265.275 112L446.61 88h33.391L330.666 200h-33.39zM480 424 330.666 312h33.391L480 398.956V424z"
      fill="#C00"
      fillRule="nonzero"
    />
  </>
);

const it = () => (
  <>
    <path fill="#fff" d="M32 88h448v336H32z" />
    <path fill="#009246" d="M32.332 88h149.336v336H32.332z" />
    <path fill="#ce2b37" d="M330.663 88H480v336H330.663z" />
  </>
);

const flags = {
  es,
  fr,
  gb,
  it,
};

export default flags;
