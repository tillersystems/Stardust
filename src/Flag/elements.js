import styled from 'styled-components';

export const Container = styled.div`
  display: inline-block;

  border-image-source: none;
  border-image-width: 0;

  padding: 0;
`;
