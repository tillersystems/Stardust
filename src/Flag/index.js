import React, { PureComponent, createRef } from 'react';
import PropTypes from 'prop-types';

import { Container } from './elements';
import flags from './data';

const { bool, string } = PropTypes;

/**
 * Flag
 *
 * This component is in charge of displaying
 * a flag with svg data
 *
 * @param {string} name // Name of the country flag. Available countries are displayed in the data file
 * @param {string} size // Size of the flag
 * @param {bool} rounded // Display the flag with rounded corners or not
 *
 * @return {jsx}
 */
class Flag extends PureComponent {
  /** Reference to the g element grouping flag svg elements. */
  groupRef = createRef();

  /**
   * PropTypes Validation.
   */
  static propTypes = {
    name: string.isRequired,
    size: string,
    rounded: bool,
  };

  /**
   * Default props.
   */
  static defaultProps = {
    size: '20',
    rounded: false,
  };

  /** Internal state. */
  state = {
    groupWidth: null,
    groupHeight: null,
    groupX: null,
    groupY: null,
  };

  /**
   * Updates the size of the rectangle that hides flag corners when rounded option is enabled
   *
   */
  componentDidMount() {
    const { rounded } = this.props;

    if (rounded) {
      const { x, y, height, width } = this.groupRef.current.getBBox();
      this.setState({ groupX: x, groupY: y, groupHeight: height, groupWidth: width });
    }
  }

  render() {
    const { name, rounded, size, ...rest } = this.props;
    const { groupX, groupY, groupHeight, groupWidth } = this.state;

    return (
      <Container {...rest}>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          style={{ display: 'block' }}
          width={size}
          height={size}
          viewBox="0 0 512 512"
          data-testid="flagSvg"
        >
          <clipPath id="rounded-corners">
            <rect rx="90" ry="90" x={groupX} y={groupY} width={groupWidth} height={groupHeight} />
          </clipPath>
          <g
            ref={this.groupRef}
            fill="none"
            fillRule="evenodd"
            clipPath={rounded ? 'url(#rounded-corners)' : undefined}
          >
            {flags[name]()}
          </g>
        </svg>
      </Container>
    );
  }
}

export default Flag;
