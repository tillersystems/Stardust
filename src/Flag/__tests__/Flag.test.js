import React from 'react';

import Flag from '..';

describe('<Flag />', () => {
  test('should render without a problem', () => {
    const { container } = render(<Flag name="fr" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render with different size', () => {
    const customSize = '50';
    const { getByTestId } = render(<Flag name="fr" size={customSize} />);
    const flagSvgNode = getByTestId('flagSvg');

    expect(flagSvgNode).toHaveAttribute('width', customSize);
    expect(flagSvgNode).toHaveAttribute('height', customSize);
  });
});
