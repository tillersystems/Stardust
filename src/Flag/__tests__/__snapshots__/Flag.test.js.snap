// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flag /> should render with rounded corners 1`] = `
.c0 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
}

<div
  class="c0"
>
  <svg
    data-testid="flagSvg"
    height="20"
    style="display: block;"
    viewBox="0 0 512 512"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <clippath
      id="rounded-corners"
    >
      <rect
        height="336"
        rx="90"
        ry="90"
        width="450"
        x="31"
        y="88"
      />
    </clippath>
    <g
      clip-path="url(#rounded-corners)"
      fill="none"
      fill-rule="evenodd"
    >
      <path
        d="M32 88h448v336H32z"
        fill="#fff"
      />
      <path
        d="M32.332 88h149.336v336H32.332z"
        fill="#00267f"
      />
      <path
        d="M330.663 88H480v336H330.663z"
        fill="#f31830"
      />
    </g>
  </svg>
</div>
`;

exports[`<Flag /> should render without a problem 1`] = `
.c0 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
}

<div
  class="c0"
>
  <svg
    data-testid="flagSvg"
    height="20"
    style="display: block;"
    viewBox="0 0 512 512"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <clippath
      id="rounded-corners"
    >
      <rect
        rx="90"
        ry="90"
      />
    </clippath>
    <g
      fill="none"
      fill-rule="evenodd"
    >
      <path
        d="M32 88h448v336H32z"
        fill="#fff"
      />
      <path
        d="M32.332 88h149.336v336H32.332z"
        fill="#00267f"
      />
      <path
        d="M330.663 88H480v336H330.663z"
        fill="#f31830"
      />
    </g>
  </svg>
</div>
`;
